propagators/opentelemetry-propagator-opentracing/src/OpenTracingPropagator.ts

```diff
@@ -0,0 +1,111 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Baggage,
+  Context,
+  getBaggage,
+  getSpanContext,
+  isSpanContextValid,
+  isValidSpanId,
+  isValidTraceId,
+  setBaggage,
+  setSpanContext,
+  TextMapGetter,
+  TextMapPropagator,
+  TextMapSetter,
+  TraceFlags,
+} from '@opentelemetry/api';
+
+/** OT header keys */
+export const OT_TRACE_ID_HEADER = 'ot-tracer-traceid';
+export const OT_SPAN_ID_HEADER = 'ot-tracer-spanid';
+export const OT_SAMPLED_HEADER = 'ot-tracer-sampled';
+export const OT_BAGGAGE_PREFIX = 'ot-tracer-baggage-';
+
+const FIELDS = [OT_TRACE_ID_HEADER, OT_SPAN_ID_HEADER, OT_SAMPLED_HEADER];
+const PADDING = '0'.repeat(16);
+
+function readHeader(
+  carrier: unknown,
+  getter: TextMapGetter,
+  key: string
+): string {
+  let header = getter.get(carrier, key);
+  if (Array.isArray(header)) [header] = header;
+  return header || '';
+}
+
+/**
+ * Propagator for the OpenTracing HTTP format.
+ */
+export class OpenTracingPropagator implements TextMapPropagator {
+  inject(context: Context, carrier: unknown, setter: TextMapSetter) {
+    const spanContext = getSpanContext(context);
+    if (!spanContext || !isSpanContextValid(spanContext)) return;
+
+    setter.set(carrier, OT_TRACE_ID_HEADER, spanContext.traceId.substr(16));
+    setter.set(carrier, OT_SPAN_ID_HEADER, spanContext.spanId);
+    setter.set(
+      carrier,
+      OT_SAMPLED_HEADER,
+      spanContext.traceFlags === TraceFlags.SAMPLED ? 'true' : 'false'
+    );
+
+    const baggage = getBaggage(context);
+    if (!baggage) return;
+
+    Object.entries(baggage).forEach(([k, v]) => {
+      setter.set(carrier, `${OT_BAGGAGE_PREFIX}${k}`, v.value);
+    });
+  }
+
+  extract(context: Context, carrier: unknown, getter: TextMapGetter): Context {
+    let traceId = readHeader(carrier, getter, OT_TRACE_ID_HEADER);
+    if (traceId.length == 16) traceId = `${PADDING}${traceId}`;
+    const spanId = readHeader(carrier, getter, OT_SPAN_ID_HEADER);
+    const sampled = readHeader(carrier, getter, OT_SAMPLED_HEADER);
+    const traceFlags =
+      sampled === 'true' ? TraceFlags.SAMPLED : TraceFlags.NONE;
+
+    if (isValidTraceId(traceId) && isValidSpanId(spanId)) {
+      context = setSpanContext(context, {
+        traceId,
+        spanId,
+        isRemote: true,
+        traceFlags,
+      });
+
+      const baggage: Baggage = getBaggage(context) || {};
+
+      getter.keys(carrier).forEach(k => {
+        if (!k.startsWith(OT_BAGGAGE_PREFIX)) return;
+        const value = readHeader(carrier, getter, k);
+        baggage[k.substr(OT_BAGGAGE_PREFIX.length)] = { value };
+      });
+
+      if (Object.keys(baggage).length > 0) {
+        context = setBaggage(context, baggage);
+      }
+    }
+
+    return context;
+  }
+
+  fields(): string[] {
+    return FIELDS;
```

Comment:
The baggage keys are missing here but spec requires them.
But honestly speaking I have no idea how to return them because they seem to depend on the context which is not available here.

Reply:
I don't think this is possible because they are per carrier instance. The jaeger format has the same issue. I took a look at the JS implementation of the Jaeger propagator, and it does not have baggage support. This is likely a bug and I can file an issue. The java implementation does support baggage, but does not include it in fields, see: https://github.com/open-telemetry/opentelemetry-java/blob/master/extensions/trace-propagators/src/main/java/io/opentelemetry/extension/trace/propagation/JaegerPropagator.java#L68.

Reply:
We should probably bring this up with spec. The usecase for fields is broken if we don't return all used fields, but saving fields that have been used could cause a memory leak.

One possible solution would be to add the list of baggage fields used to the context returned by `inject`.

Reply:
Avoid letting the caller mutate our fields list

```suggestion
    return FIELDS.slice();
```